### ISSUE TITLE
when a job exceeds its time limit, fluxion may reallocate resources before they have been freed

### ISSUE THREAD

**Issue by garlick** [flux-framework/flux-sched#1043]
Problem: when scheduling back to back jobs using Fluxion in `lonodex` mode with prolog and epilog scripts each containing a 10s sleep, I observed prolog and epilog scripts executing simultaneously, which seems like it might cause problems?

Here's a snapshot of `flux jobs` showing the effect:
```
 garlick@picl0:~$ flux jobs
       JOBID QUEUE    USER     NAME       ST NTASKS NNODES     TIME INFO
 ƒTF5edtS4D5 debug    garlick  stress      R      1      1   2.167s picl1
 ƒTF54wFgwgT debug    garlick  stress      C      1      1   1.177m picl1
```

This is on my `sdexec` working branch so it's possible (though I'm not seeing how) that I introduced something bad:
```
commands:    		0.51.0-160-g34e483a82
libflux-core:		0.51.0-160-g34e483a82
libflux-security:	0.9.0
build-options:		+systemd+hwloc==2.4.0+zmq==4.3.4

sched-fluxion-qmanager.info[0]: version 0.25.0-12-gecf20b37
```

**Comment by grondo** [comment]
tagging @trws, @milroy or maybe @jameshcorbett for an assist on this one. This could be considered a critical issue if it is occurring in production (likely, since all jobs have time limits by default)

**Comment by garlick** [comment]
A guess is that `fluxion-resource` is making resources available again once their allocation's time limit has been reached, rather than waiting for them to be explicitly freed.  Since the epilog in this case runs after the job has exceeded its time limit but before resources are freed, we have sad trombones. :sob: :trumpet:  :bone:

**Comment by garlick** [comment]
For a reliable reproducer, see flux-framework/flux-core#5304.

**Comment by milroy** [comment]
I can produce the reported behavior via the reproducer in flux-core #5304.

I can also reproduce the behavior with various `qmanager` options, e.g., running with `queue-policy=fcfs queue-params=queue-depth=1`. If I increase the job time limit to 3-4s the behavior stops, suggesting that this is a race condition.

Due to the way Fluxion checks temporal availability of resources in its DFS, if the time at which it queries a vertex's planner is after the previous allocation's walltime (i.e., the scheduler's `at` time is after the planner's earliest available time), Fluxion considers that vertex eligible for allocation (even if its vertex-specific map is nonempty). I think no one observed this behavior before because submitting a large number of very short walltime jobs at once on a system with a very small resource graph in comparison to the resource request dramatically increases the likelihood of a traversal happening after a vertex becomes available but before the resources are freed.

**Comment by milroy** [comment]
I agree with the assessment that the issue is critical and will continue to work on it.

**Comment by milroy** [comment]
@garlick: I don't understand why `submit` with `--cc` avoids this problem. Does that make sense to you?

**Comment by milroy** [comment]
Are there any constraints on the contents of an epilog script or limits to how long it can run? I suspect this problem will also occur if a job gets stuck in cleanup for a long time.

**Comment by grondo** [comment]
The design allows the epilog to run for infinite time. This could occur if a filesystem is hung or some other node problem causes a long delay in job cleanup. The scheduler absolutely cannot release resources for a job until a `free` request is made.

**Comment by garlick** [comment]
>  I don't understand why submit with --cc avoids this problem. Does that make sense to you?

I'm not sure why that might be.  My first thought was that, when submitted together, the jobs might be handled by one scheduling loop as opposed to triggering a loop at each submission, but I didn't confirm that.

If you're now basing your testing on #1044, this diff demonstrates the effect
```patch
diff --git a/t/t1024-alloc-check.t b/t/t1024-alloc-check.t
index ce5fa2a5..6fc4b41c 100755
--- a/t/t1024-alloc-check.t
+++ b/t/t1024-alloc-check.t
@@ -24,9 +24,10 @@ test_expect_success 'load alloc-check plugin' '
 '
 # Jobs seem to need to be submitted separately to trigger the issue.
 test_expect_success 'submit consecutive jobs that exceed their time limit' '
-       (for i in $(seq 5); do \
-           flux run -N1 -x -t1s sleep 30 || true; \
-       done) 2>joberr
+       flux submit --cc 1-5 -N1 -x -t1s sleep 30 1>jobids &&
+       for id in $(cat jobids); do \
+           (flux job attach $id || true) 2>joberr; \
+       done
 '
 test_expect_success 'some jobs received timeout exception' '
        grep "job.exception type=timeout" joberr
```

**Comment by milroy** [comment]
> I should add the schedule Fluxion produces is based on the assumption that the walltime and the end of the job are the same. 
I should correct this statement. Fluxion makes scheduling decisions based on the assumption that the end of the job will never be later than the walltime. This issue demonstrates the negative impact of that incorrect assumption.

**Comment by milroy** [comment]
After my conversation with @garlick, I pondered several schemes for addressing this issue. Many are quite complex and prone to race conditions.

I thought more about the idea (that Jim suggested yesterday and I tried on Wednesday) of just constraining allocations to occur on exclusive vertices when there are no existing allocations. I realized my previous implementation was wrong and corrected it. I think this simple change to `resource/traversers/dfu_impl.cpp` fixes the issue in the reproducer:

```diff
diff --git a/resource/traversers/dfu_impl.cpp b/resource/traversers/dfu_impl.cpp
index 39f8e7f0..06b0f99b 100644
--- a/resource/traversers/dfu_impl.cpp
+++ b/resource/traversers/dfu_impl.cpp
@@ -125,6 +125,8 @@ int dfu_impl_t::by_excl (const jobmeta_t &meta, const std::string &s, vtx_t u,
     // its x_checker planner.
     if (exclusive_in || resource.exclusive == Jobspec::tristate_t::TRUE) {
         errno = 0;
+	if (meta.alloc_type == jobmeta_t::alloc_type_t::AT_ALLOC && !(*m_graph)[u].schedule.allocations.empty ())
+	    goto restore_errno;
         p = (*m_graph)[u].idata.x_checker;
         njobs = planner_avail_resources_during (p, at, duration);
         if (njobs == -1) {
```
@garlick and @grondo please give that a try when you get a chance.

Note that I'm not convinced this is a general solution, but I think it's on the right track and does not appear to have side effects. It also passes Fluxion testsuite.

**Comment by garlick** [comment]
I confirmed that with this change, the test posted in #1044 now passes and the rest of the test suite passes for me as well.

I'm running the reproducer in a loop now - so far so good!

I'll also install on my test cluster and throw random workloads at it with the alloc-check plugin loaded.

**Comment by milroy** [comment]
This is good news!

> I'll also install on my test cluster and throw random workloads at it with the alloc-check plugin loaded.

That's a great idea. I'm particularly concerned about how the fix handles non-exclusive resource allocations.

**Comment by garlick** [comment]
FWIW: just pushed a test to #1044 that runs the same test with non-exclusively scheduled nodes.   Seems to fail the same as for exclusive nodes without the proposed fix, and not fail with the proposed fix.

**Comment by milroy** [comment]
Thanks for the additional testing. I'll create a WIP PR for the fix this afternoon.

**Comment by garlick** [comment]
Sounds good and maybe you can have a go at explaining the fix to me and @grondo on the 2pm coffee call.  You said I suggested this but I suggested generalities (mostly ignorant of fluxion internals0 and you figured it out :-)   Also, it would be good to know why you said this is not the general solution, and what the general solution might be.

**Comment by milroy** [comment]
I opened PR #1046 that has a more detailed explanation of what the commit does. Note that I slightly changed the commit to `goto` before setting `errno` which should behave the same.

**Comment by grondo** [comment]
Should this have been closed by #1046?
